File: sdk/elixir/lib/rocksky/player.ex
defmodule Rocksky.Player do
  @moduledoc "`app.rocksky.player.*` endpoints — remote control of a Rocksky player."

  alias Rocksky.HTTP

  @doc "Currently playing track. Params: `:playerId`, `:actor`."
  def get_currently_playing(client, params \\ []),
    do: HTTP.query(client, "app.rocksky.player.getCurrentlyPlaying", params)

  @doc "Playback queue. Params: `:playerId`."
  def get_playback_queue(client, params \\ []),
    do: HTTP.query(client, "app.rocksky.player.getPlaybackQueue", params)

  @doc "Resume playback. Params: `:playerId`."
  def play(client, params \\ []),
    do: HTTP.procedure(client, "app.rocksky.player.play", params)

  @doc "Pause playback. Params: `:playerId`."
  def pause(client, params \\ []),
    do: HTTP.procedure(client, "app.rocksky.player.pause", params)

  @doc "Skip to next track. Params: `:playerId`."
  def next(client, params \\ []),
    do: HTTP.procedure(client, "app.rocksky.player.next", params)

  @doc "Go to previous track. Params: `:playerId`."
  def previous(client, params \\ []),
    do: HTTP.procedure(client, "app.rocksky.player.previous", params)

  @doc "Seek to position (ms). Params: `:playerId`, `:position`."
  def seek(client, params \\ []),
    do: HTTP.procedure(client, "app.rocksky.player.seek", params)

  @doc "Play a single file. Params: `:playerId`, `:fileId`."
  def play_file(client, params \\ []),
    do: HTTP.procedure(client, "app.rocksky.player.playFile", params)

  @doc "Play a directory. Params: `:playerId`, `:directoryId`, `:shuffle`, `:recurse`, `:position`."
  def play_directory(client, params \\ []),
    do: HTTP.procedure(client, "app.rocksky.player.playDirectory", params)

  @doc "Append items to the queue. Params: `:playerId`, `:items`, `:position`, `:shuffle`."
  def add_items_to_queue(client, params \\ []),
    do: HTTP.procedure(client, "app.rocksky.player.addItemsToQueue", params)
end

File: sdk/clojure/src/rocksky/player.clj
(ns rocksky.player
  "Endpoints under app.rocksky.player.* — controls Rocksky's web player."
  (:require [rocksky.client :as c]))

(defn get-currently-playing
  "Get the currently playing track.

  Optional: `:player-id` `:actor`."
  ([client] (get-currently-playing client nil))
  ([client {:keys [player-id actor]}]
   (c/query client :app.rocksky.player.getCurrentlyPlaying
            {:playerId player-id :actor actor})))

(defn get-playback-queue
  "Get the playback queue. Optional: `:player-id`."
  ([client] (get-playback-queue client nil))
  ([client {:keys [player-id]}]
   (c/query client :app.rocksky.player.getPlaybackQueue
            {:playerId player-id})))

(defn play
  "Resume playback. Optional: `:player-id`."
  ([client] (play client nil))
  ([client {:keys [player-id]}]
   (c/procedure client :app.rocksky.player.play nil
                {:playerId player-id})))

(defn pause
  "Pause playback. Optional: `:player-id`."
  ([client] (pause client nil))
  ([client {:keys [player-id]}]
   (c/procedure client :app.rocksky.player.pause nil
                {:playerId player-id})))

(defn next-track
  "Skip to the next track. Optional: `:player-id`."
  ([client] (next-track client nil))
  ([client {:keys [player-id]}]
   (c/procedure client :app.rocksky.player.next nil
                {:playerId player-id})))

(defn previous-track
  "Skip to the previous track. Optional: `:player-id`."
  ([client] (previous-track client nil))
  ([client {:keys [player-id]}]
   (c/procedure client :app.rocksky.player.previous nil
                {:playerId player-id})))

(defn seek
  "Seek to `:position` (in seconds). Required: `:position`. Optional: `:player-id`."
  [client {:keys [player-id position]}]
  (c/procedure client :app.rocksky.player.seek nil
               {:playerId player-id :position position}))

(defn add-items-to-queue
  "Add items to the queue.

  Required: `:items` (collection of file IDs).
  Optional: `:player-id` `:position` `:shuffle`."
  [client {:keys [player-id items position shuffle]}]
  (c/procedure client :app.rocksky.player.addItemsToQueue nil
               {:playerId player-id
                :items    items
                :position position
                :shuffle  shuffle}))

(defn play-file
  "Play a specific file by `:file-id`.

  Required: `:file-id`. Optional: `:player-id`."
  [client {:keys [player-id file-id]}]
  (c/procedure client :app.rocksky.player.playFile nil
               {:playerId player-id :fileId file-id}))

(defn play-directory
  "Play all tracks in a directory.

  Required: `:directory-id`.
  Optional: `:player-id` `:shuffle` `:recurse` `:position`."
  [client {:keys [player-id directory-id shuffle recurse position]}]
  (c/procedure client :app.rocksky.player.playDirectory nil
               {:playerId    player-id
                :directoryId directory-id
                :shuffle     shuffle
                :recurse     recurse
                :position    position}))
